// @fluentui/react-northstar/dist/es/components/Datepicker/DatepickerCalendarHeaderAction.js
import * as React from 'react';
import { compose } from '@fluentui/react-bindings';
import { ChevronEndIcon, ChevronStartIcon } from '@fluentui/react-icons-northstar';
import { Button } from '../Button/Button';
export var datepickerCalendarHeaderActionClassName = 'ui-datepicker__calendarheaderaction';
/**
 * A DatepickerCalendarHeaderAction is used to display action button for DatepickerCalendarHeader.
 * This component is currently UNSTABLE!
 */

export var DatepickerCalendarHeaderAction = compose(Button, {
  className: datepickerCalendarHeaderActionClassName,
  displayName: 'DatepickerCalendarHeaderAction',
  handledProps: ['direction'],
  mapPropsToStylesProps: function mapPropsToStylesProps() {
    return {
      iconOnly: true,
      text: true
    };
  },
  slotProps: function slotProps(props) {
    return {
      icon: {
        content: props.direction === 'next' ? /*#__PURE__*/React.createElement(ChevronEndIcon, null) : /*#__PURE__*/React.createElement(ChevronStartIcon, null)
      }
    };
  }
});
