// @pnp/spfx-controls-react/lib/controls/dynamicForm/dynamicField/DynamicField.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import * as React from 'react';
import styles from '../DynamicForm.module.scss';
import { Dropdown } from 'office-ui-fabric-react/lib/components/Dropdown';
import { DatePicker } from 'office-ui-fabric-react/lib/DatePicker';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { PeoplePicker, PrincipalType } from '../../peoplepicker';
import { FilePicker } from '../../filePicker';
import { TaxonomyPicker } from '../../taxonomyPicker';
import { ListItemPicker } from '../../listItemPicker';
import { LocationPicker } from '../../locationPicker';
import { RichText } from '../../richText';
import { Icon } from 'office-ui-fabric-react/lib/Icon';
import { Shimmer } from 'office-ui-fabric-react/lib/Shimmer';
import { Toggle } from 'office-ui-fabric-react/lib/Toggle';
import { Image } from 'office-ui-fabric-react/lib/Image';
import { Stack } from 'office-ui-fabric-react/lib/Stack';
import { DateTimePicker } from '../../dateTimePicker/DateTimePicker';
import { sp } from '@pnp/sp/presets/all';
import * as strings from 'ControlStrings';
import '@pnp/sp/folders';
import '@pnp/sp/webs';
import { ActionButton } from 'office-ui-fabric-react';
var DynamicField = /** @class */ (function (_super) {
    __extends(DynamicField, _super);
    function DynamicField(props) {
        var _this = _super.call(this, props) || this;
        _this.getFieldComponent = function () {
            var _a = _this.props, options = _a.options, fieldTermSetId = _a.fieldTermSetId, lookupListID = _a.lookupListID, lookupField = _a.lookupField, fieldType = _a.fieldType, fieldDefaultValue = _a.fieldDefaultValue, fieldTitle = _a.fieldTitle, context = _a.context, disabled = _a.disabled, label = _a.label, placeholder = _a.placeholder, required = _a.required, isRichText = _a.isRichText, 
            //bingAPIKey,
            dateFormat = _a.dateFormat, columnInternalName = _a.columnInternalName, principalType = _a.principalType, description = _a.description;
            var changedValue = _this.state.changedValue;
            var dropdownOptions = {
                options: options,
                disabled: disabled,
                placeholder: placeholder
            };
            var labelText = fieldTitle != null ? fieldTitle : label;
            var defaultValue = fieldDefaultValue;
            var empty = null;
            var labelEl = React.createElement("label", { className: (required) ? styles.fieldRequired + ' ' + styles.fieldLabel : styles.fieldLabel }, labelText);
            var errorText = _this.getRequiredErrorText();
            var errorTextEl = React.createElement("text", { className: styles.errormessage }, errorText);
            var descriptionEl = React.createElement("text", { className: styles.fieldDescription }, description);
            switch (fieldType) {
                case 'loading':
                    return React.createElement(Shimmer, { width: "75%", styles: {
                            root: {
                                margin: '25px'
                            }
                        } });
                case 'Text':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "TextField" }),
                            labelEl),
                        React.createElement(TextField, { defaultValue: defaultValue, placeholder: placeholder, className: styles.feildDisplay, onChange: function (e, newText) { _this.onChange(newText); }, disabled: disabled, onBlur: _this.onBlur, errorMessage: errorText }),
                        descriptionEl);
                case 'Note':
                    if (isRichText) {
                        return React.createElement("div", { className: styles.richText },
                            React.createElement("div", { className: styles.titleContainer },
                                React.createElement(Icon, { className: styles.fieldIcon, iconName: "AlignLeft" }),
                                labelEl),
                            React.createElement(RichText, { placeholder: placeholder, value: defaultValue, className: styles.feildDisplay, onChange: function (newText) { _this.onChange(newText); return newText; }, isEditMode: disabled }),
                            descriptionEl,
                            errorTextEl);
                    }
                    else {
                        return React.createElement("div", null,
                            React.createElement("div", { className: styles.titleContainer },
                                React.createElement(Icon, { className: styles.fieldIcon, iconName: "AlignLeft" }),
                                labelEl),
                            React.createElement(TextField, { defaultValue: defaultValue, placeholder: placeholder, className: styles.feildDisplay, multiline: true, onChange: function (e, newText) { _this.onChange(newText); }, disabled: disabled, onBlur: _this.onBlur, errorMessage: errorText }),
                            descriptionEl);
                    }
                case 'Choice':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.labelContainer + " " + styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "CheckMark" }),
                            labelEl),
                        React.createElement(Dropdown, __assign({}, dropdownOptions, { defaultSelectedKey: defaultValue, onChange: function (e, option) { _this.onChange(option); }, onBlur: _this.onBlur, errorMessage: errorText })),
                        descriptionEl);
                case 'MultiChoice':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.labelContainer + " " + styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "MultiSelect" }),
                            labelEl),
                        React.createElement(Dropdown, __assign({}, dropdownOptions, { defaultSelectedKeys: defaultValue, onChange: _this.MultiChoice_selection, multiSelect: true, onBlur: _this.onBlur, errorMessage: errorText })),
                        descriptionEl);
                case 'Location':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.labelContainer + " " + styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "POI" }),
                            labelEl),
                        React.createElement(LocationPicker, { context: context, disabled: disabled, placeholder: placeholder, onChange: function (newValue) { _this.onChange(newValue); }, defaultValue: defaultValue, errorMessage: errorText }),
                        descriptionEl);
                case 'Lookup':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Switch" }),
                            labelEl),
                        React.createElement(ListItemPicker, { disabled: disabled, listId: lookupListID, defaultSelectedItems: defaultValue, columnInternalName: lookupField, className: styles.feildDisplay, keyColumnInternalName: 'Id', itemLimit: 1, onSelectedItem: function (newValue) { _this.onChange(newValue); }, context: context }),
                        descriptionEl,
                        errorTextEl);
                case 'LookupMulti':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Switch" }),
                            labelEl),
                        React.createElement(ListItemPicker, { disabled: disabled, listId: lookupListID, defaultSelectedItems: defaultValue, columnInternalName: lookupField, className: styles.feildDisplay, keyColumnInternalName: 'Id', itemLimit: 100, onSelectedItem: function (newValue) { _this.onChange(newValue); }, context: context }),
                        descriptionEl,
                        errorTextEl);
                case 'Number':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "NumberField" }),
                            labelEl),
                        React.createElement(TextField, { defaultValue: defaultValue, placeholder: placeholder, className: styles.feildDisplay, type: "Number", onChange: function (e, newText) { _this.onChange(newText); }, disabled: disabled, onBlur: _this.onBlur, errorMessage: errorText }),
                        descriptionEl);
                case 'Currency':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "AllCurrency" }),
                            labelEl),
                        React.createElement(TextField, { defaultValue: defaultValue, placeholder: placeholder, className: styles.feildDisplay, type: "Currency", onChange: function (e, newText) { _this.onChange(newText); }, disabled: disabled, onBlur: _this.onBlur, errorMessage: errorText }),
                        descriptionEl);
                case 'DateTime':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Calendar" }),
                            labelEl),
                        dateFormat === 'DateOnly' &&
                            React.createElement(DatePicker, { placeholder: placeholder, className: styles.pickersContainer, formatDate: function (date) { return date.toLocaleDateString(context.pageContext.web.languageName); }, value: (changedValue !== null && changedValue !== "") ? changedValue : defaultValue, onSelectDate: function (newDate) { _this.onChange(newDate); }, disabled: disabled }),
                        dateFormat === 'DateTime' &&
                            React.createElement(DateTimePicker, { key: columnInternalName, placeholder: placeholder, formatDate: function (date) { return date.toLocaleDateString(context.pageContext.web.languageName); }, value: (changedValue !== null && changedValue !== "") ? changedValue : defaultValue, onChange: function (newDate) { _this.onChange(newDate); }, disabled: disabled }),
                        descriptionEl,
                        errorTextEl);
                case 'Boolean':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "CheckboxComposite" }),
                            labelEl),
                        React.createElement(Toggle, { className: styles.feildDisplay, defaultChecked: defaultValue, onText: strings.Yes, offText: strings.No, onChange: function (e, checkedvalue) { _this.onChange(checkedvalue); }, disabled: disabled }),
                        descriptionEl,
                        errorTextEl);
                case 'User':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Contact" }),
                            labelEl),
                        React.createElement(PeoplePicker, { placeholder: placeholder, defaultSelectedUsers: defaultValue, peoplePickerCntrlclassName: styles.feildDisplay, context: context, personSelectionLimit: 1, showtooltip: false, showHiddenInUI: false, principalTypes: principalType === 'PeopleOnly' ? [PrincipalType.User] : [PrincipalType.User, PrincipalType.SharePointGroup, PrincipalType.DistributionList, PrincipalType.SecurityGroup], resolveDelay: 1000, onChange: function (items) { _this.onChange(items); }, disabled: disabled }),
                        descriptionEl,
                        errorTextEl);
                case 'UserMulti':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Contact" }),
                            labelEl),
                        React.createElement(PeoplePicker, { placeholder: placeholder, defaultSelectedUsers: defaultValue, peoplePickerCntrlclassName: styles.feildDisplay, context: context, personSelectionLimit: 30, showtooltip: false, showHiddenInUI: false, principalTypes: principalType === 'PeopleOnly' ? [PrincipalType.User] : [PrincipalType.User, PrincipalType.SharePointGroup, PrincipalType.DistributionList, PrincipalType.SecurityGroup], resolveDelay: 1000, onChange: function (items) { _this.onChange(items); }, disabled: disabled }),
                        descriptionEl,
                        errorTextEl);
                case 'URL':
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "Link" }),
                            labelEl),
                        React.createElement(Stack, { tokens: { childrenGap: 4 } },
                            React.createElement(TextField, { defaultValue: defaultValue ? defaultValue['Url'] : '', placeholder: strings.DynamicFormEnterURLPlaceholder, className: styles.feildDisplayNoPadding, onChange: function (e, newText) { _this.onURLChange(newText, true); }, disabled: disabled, onBlur: _this.onBlur }),
                            React.createElement(TextField, { defaultValue: defaultValue ? defaultValue['Description'] : '', placeholder: strings.DynamicFormEnterDescriptionPlaceholder, className: styles.feildDisplayNoPadding, onChange: function (e, newText) { _this.onURLChange(newText, false); }, disabled: disabled })),
                        descriptionEl,
                        errorTextEl);
                case 'Thumbnail':
                    var hasImage = !!changedValue; // || !!defaultValue;
                    return React.createElement("div", null,
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "photo2" }),
                            labelEl),
                        React.createElement(Stack
                        //className={styles.filePicker}
                        , { 
                            //className={styles.filePicker}
                            horizontal: true, tokens: {
                                childrenGap: 20
                            }, horizontalAlign: 'space-between' },
                            hasImage && React.createElement(Image, { src: changedValue, height: 60 }),
                            React.createElement("div", { className: styles.thumbnailFieldButtons },
                                React.createElement(FilePicker, { buttonClassName: styles.feildDisplay, 
                                    //bingAPIKey={bingAPIKey}
                                    accepts: [".gif", ".jpg", ".jpeg", ".bmp", ".dib", ".tif", ".tiff", ".ico", ".png", ".jxr", ".svg"], buttonLabel: hasImage ? undefined : 'Add an image', buttonIcon: hasImage ? 'Edit' : 'FileImage', onSave: _this.saveIntoSharePoint, onChange: _this.saveIntoSharePoint, context: context, disabled: disabled, hideLocalMultipleUploadTab: true, hideOneDriveTab: true, hideStockImages: true, hideWebSearchTab: true }),
                                hasImage &&
                                    React.createElement(ActionButton, { disabled: disabled, iconProps: {
                                            iconName: 'Delete'
                                        }, onClick: _this.onDeleteImage }))),
                        descriptionEl,
                        errorTextEl);
                case 'TaxonomyFieldTypeMulti':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "BulletedTreeList" }),
                            labelEl),
                        React.createElement("div", { className: styles.pickersContainer },
                            React.createElement(TaxonomyPicker, { label: "", disabled: disabled, initialValues: defaultValue, placeholder: placeholder, allowMultipleSelections: true, termsetNameOrID: fieldTermSetId, panelTitle: strings.DynamicFormTermPanelTitle, context: context, onChange: function (newValue) { _this.onChange(newValue); }, isTermSetSelectable: false })),
                        descriptionEl,
                        errorTextEl);
                case 'TaxonomyFieldType':
                    return React.createElement("div", { className: styles.fieldContainer },
                        React.createElement("div", { className: styles.titleContainer },
                            React.createElement(Icon, { className: styles.fieldIcon, iconName: "BulletedTreeList" }),
                            labelEl),
                        React.createElement("div", { className: styles.pickersContainer },
                            React.createElement(TaxonomyPicker, { label: "", disabled: disabled, initialValues: defaultValue, placeholder: placeholder, allowMultipleSelections: false, termsetNameOrID: fieldTermSetId, panelTitle: strings.DynamicFormTermPanelTitle, context: context, onChange: function (newValue) { _this.onChange(newValue); }, isTermSetSelectable: false })),
                        descriptionEl,
                        errorTextEl);
            }
            return null;
        };
        _this.onDeleteImage = function () {
            var _a = _this.props, onChanged = _a.onChanged, columnInternalName = _a.columnInternalName;
            _this.setState({
                changedValue: undefined
            });
            if (onChanged) {
                onChanged(columnInternalName, undefined, undefined);
            }
        };
        _this.onURLChange = function (value, isUrl) {
            var _a = _this.props, fieldDefaultValue = _a.fieldDefaultValue, onChanged = _a.onChanged, columnInternalName = _a.columnInternalName;
            var currValue = _this.state.changedValue || fieldDefaultValue || {
                Url: '',
                Description: ''
            };
            currValue = __assign({}, currValue);
            if (isUrl) {
                currValue.Url = value;
            }
            else {
                currValue.Description = value;
            }
            _this.setState({
                changedValue: currValue
            });
            if (onChanged) {
                onChanged(columnInternalName, currValue);
            }
        };
        _this.onChange = function (value) {
            var _a = _this.props, onChanged = _a.onChanged, columnInternalName = _a.columnInternalName;
            if (onChanged) {
                onChanged(columnInternalName, value);
            }
            _this.setState({
                changedValue: value
            });
        };
        _this.onBlur = function () {
            if (_this.state.changedValue === null && _this.props.fieldDefaultValue === "") {
                _this.setState({ changedValue: "" });
            }
        };
        _this.getRequiredErrorText = function () {
            var changedValue = _this.state.changedValue;
            return (changedValue === undefined || changedValue === '') && _this.props.required ? strings.DynamicFormRequiredErrorMessage : null;
        };
        _this.MultiChoice_selection = function (event, item) {
            var changedValue = _this.state.changedValue;
            try {
                var seletedItemArr_1;
                if (changedValue === null && _this.props.fieldDefaultValue != null) {
                    seletedItemArr_1 = [];
                    _this.props.fieldDefaultValue.forEach(function (element) {
                        seletedItemArr_1.push(element);
                    });
                }
                else
                    seletedItemArr_1 = !changedValue ? [] : changedValue;
                if (item.selected) {
                    seletedItemArr_1.push(item.key);
                }
                else {
                    var i = seletedItemArr_1.indexOf(item.key);
                    if (i >= 0) {
                        seletedItemArr_1.splice(i, 1);
                    }
                }
                _this.setState({ changedValue: seletedItemArr_1 });
                _this.props.onChanged(_this.props.columnInternalName, seletedItemArr_1);
            }
            catch (error) {
                console.log("Error MultiChoice_selection", error);
            }
        };
        _this.saveIntoSharePoint = function (files) { return __awaiter(_this, void 0, void 0, function () {
            var _a, columnInternalName, onChanged, newValue, file;
            return __generator(this, function (_b) {
                _a = this.props, columnInternalName = _a.columnInternalName, onChanged = _a.onChanged;
                if (!files.length) {
                    return [2 /*return*/];
                }
                try {
                    file = files[0];
                    if (file.fileAbsoluteUrl === null) {
                        newValue = file.previewDataUrl;
                    }
                    else {
                        newValue = file.fileAbsoluteUrl;
                    }
                    this.setState({
                        changedValue: newValue
                    });
                    if (onChanged) {
                        onChanged(columnInternalName, newValue, file);
                    }
                }
                catch (error) {
                    console.log("Error save Into SharePoint", error);
                }
                return [2 /*return*/];
            });
        }); };
        sp.setup({
            spfxContext: _this.props.context
        });
        _this.state = {
            changedValue: props.fieldType === 'Thumbnail' ? props.fieldDefaultValue : null
        };
        return _this;
    }
    DynamicField.prototype.componentDidUpdate = function () {
        if (this.props.fieldDefaultValue === "" && this.state.changedValue === null) {
            this.setState({ changedValue: "" });
        }
    };
    DynamicField.prototype.render = function () {
        try {
            return (React.createElement("div", { className: styles.FieldEditor }, this.getFieldComponent()));
        }
        catch (error) {
            console.log("Error in DynamicField render", error);
            return null;
        }
    };
    return DynamicField;
}(React.Component));
export { DynamicField };
