// @pnp/spfx-controls-react/lib/controls/richText/RteColorPicker.module.scss.js
/* tslint:disable */
require("./RteColorPicker.module.css");
var styles = {
    ThemeColorDarker: '#004578',
    ThemeColorDark: '#005a9e',
    ThemeColorDarkAlt: '#106ebe',
    ThemeColorPrimary: '#0078d4',
    ThemeColorSecondary: '#2b88d8',
    ThemeColorTertiary: '#71afe5',
    ThemeColorNeutralSecondary: '#605e5c',
    ThemeColorNeutralPrimaryAlt: '#3b3a39',
    ThemeColorNeutralPrimary: '#323130',
    ThemeColorNeutralDark: '#201f1e',
    colorPickerButton: 'colorPickerButton_3aee6810',
    previewSvg: 'previewSvg_3aee6810',
    border: 'border_3aee6810',
    previewRectangle: 'previewRectangle_3aee6810',
    buttonLabel: 'buttonLabel_3aee6810',
    previewIcon: 'previewIcon_3aee6810',
    pickerCallout: 'pickerCallout_3aee6810',
    defaultSvg: 'defaultSvg_3aee6810',
    fillThemeColor: 'fillThemeColor_3aee6810',
    fillDefaultColor: 'fillDefaultColor_3aee6810',
    pickerLabel: 'pickerLabel_3aee6810',
    focusedContainer: 'focusedContainer_3aee6810',
    tableRoot: 'tableRoot_3aee6810',
    tableCell: 'tableCell_3aee6810',
    colorCell: 'colorCell_3aee6810',
    svg: 'svg_3aee6810',
    selected: 'selected_3aee6810'
};
export default styles;
/* tslint:enable */
