// @pnp/spfx-controls-react/lib/controls/carousel/Carousel.module.scss.js
/* tslint:disable */
require("./Carousel.module.css");
var styles = {
    root: 'root_14843c4a',
    indicatorsContainer: 'indicatorsContainer_14843c4a',
    indicators: 'indicators_14843c4a',
    container: 'container_14843c4a',
    contentContainer: 'contentContainer_14843c4a',
    rectangle: 'rectangle_14843c4a',
    square: 'square_14843c4a',
    circle: 'circle_14843c4a',
    active: 'active_14843c4a',
    slideWrapper: 'slideWrapper_14843c4a',
    left: 'left_14843c4a',
    right: 'right_14843c4a',
    slideLeft: 'slideLeft_14843c4a',
    slideleft: 'slideleft_14843c4a',
    slideRight: 'slideRight_14843c4a',
    slideright: 'slideright_14843c4a',
    slideFromRight: 'slideFromRight_14843c4a',
    slidefromright: 'slidefromright_14843c4a',
    slideFromLeft: 'slideFromLeft_14843c4a',
    slidefromleft: 'slidefromleft_14843c4a',
    carouselImage: 'carouselImage_14843c4a',
    image: 'image_14843c4a',
    staticDetails: 'staticDetails_14843c4a',
    details: 'details_14843c4a',
    title: 'title_14843c4a',
    dynamicDetails: 'dynamicDetails_14843c4a',
    loadingComponent: 'loadingComponent_14843c4a',
    buttonLocations: 'buttonLocations_14843c4a',
    centralButtonsContainer: 'centralButtonsContainer_14843c4a',
    topButtonsContainer: 'topButtonsContainer_14843c4a',
    bottomButtonsContainer: 'bottomButtonsContainer_14843c4a',
    buttonsOnlyPrevButton: 'buttonsOnlyPrevButton_14843c4a',
    buttonsOnlyNextButton: 'buttonsOnlyNextButton_14843c4a',
    buttonsContainer: 'buttonsContainer_14843c4a',
    buttonsOnlyContainer: 'buttonsOnlyContainer_14843c4a',
    blockButtonsContainer: 'blockButtonsContainer_14843c4a',
    hiddenButtonsContainer: 'hiddenButtonsContainer_14843c4a'
};
export default styles;
/* tslint:enable */
