// @pnp/spfx-controls-react/lib/controls/filePicker/MultipleUploadFilePickerTab/MultipleUploadFilePickerTab.module.scss.js
/* tslint:disable */
require("./MultipleUploadFilePickerTab.module.css");
var styles = {
    filePicker: 'filePicker_dc9584b7',
    tabsContainer: 'tabsContainer_dc9584b7',
    tabContainer: 'tabContainer_dc9584b7',
    nav: 'nav_dc9584b7',
    tabHeaderContainer: 'tabHeaderContainer_dc9584b7',
    breadcrumbNav: 'breadcrumbNav_dc9584b7',
    breadcrumbNavItem: 'breadcrumbNavItem_dc9584b7',
    tab: 'tab_dc9584b7',
    tabOffset: 'tabOffset_dc9584b7',
    tabFiles: 'tabFiles_dc9584b7',
    scrollablePaneWrapper: 'scrollablePaneWrapper_dc9584b7',
    actionButtonsContainer: 'actionButtonsContainer_dc9584b7',
    actionButtons: 'actionButtons_dc9584b7',
    actionButton: 'actionButton_dc9584b7',
    tabHeader: 'tabHeader_dc9584b7',
    selectedTab: 'selectedTab_dc9584b7',
    header: 'header_dc9584b7',
    xlg: 'xlg_dc9584b7',
    xxlg: 'xxlg_dc9584b7',
    xxxlg: 'xxxlg_dc9584b7',
    xxxxlg: 'xxxxlg_dc9584b7',
    itemPickerTopBar: 'itemPickerTopBar_dc9584b7',
    commandBarNoChevron: 'commandBarNoChevron_dc9584b7',
    fileItem: 'fileItem_dc9584b7',
    folderItem: 'folderItem_dc9584b7',
    fileTypeIcon: 'fileTypeIcon_dc9584b7',
    fileTypeIconIcon: 'fileTypeIconIcon_dc9584b7',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_dc9584b7',
    iconColumnHeader: 'iconColumnHeader_dc9584b7',
    fileRow: 'fileRow_dc9584b7',
    folderRow: 'folderRow_dc9584b7',
    emptyFolder: 'emptyFolder_dc9584b7',
    emptyFolderImage: 'emptyFolderImage_dc9584b7',
    emptyFolderImageTag: 'emptyFolderImageTag_dc9584b7',
    emptyFolderTitle: 'emptyFolderTitle_dc9584b7',
    emptyFolderSubText: 'emptyFolderSubText_dc9584b7',
    emptyFolderMobile: 'emptyFolderMobile_dc9584b7',
    emptyFolderPc: 'emptyFolderPc_dc9584b7',
    emptyFolderIcon: 'emptyFolderIcon_dc9584b7',
    localTabSinglePreview: 'localTabSinglePreview_dc9584b7',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_dc9584b7',
    localTabFilename: 'localTabFilename_dc9584b7',
    localTabLabel: 'localTabLabel_dc9584b7',
    localTabInput: 'localTabInput_dc9584b7',
    localTabDragDropFile: 'localTabDragDropFile_dc9584b7'
};
export default styles;
/* tslint:enable */
