// @fluentui/react-northstar/dist/commonjs/components/Datepicker/DatepickerCalendarHeaderCell.js
"use strict";

exports.__esModule = true;
exports.DatepickerCalendarHeaderCell = exports.datepickerCalendarHeaderCellClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var datepickerCalendarHeaderCellClassName = 'ui-datepicker__calendarheadercell';
/**
 * A DatepickerCalendarHeaderCell is used to display header cells in DatepickerCalendar grid.
 * This component is currently UNSTABLE!
 */

exports.datepickerCalendarHeaderCellClassName = datepickerCalendarHeaderCellClassName;
var DatepickerCalendarHeaderCell = (0, _reactBindings.compose)(_Box.Box, {
  className: datepickerCalendarHeaderCellClassName,
  displayName: 'DatepickerCalendarHeaderCell',
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.DatepickerCalendarHeaderCell = DatepickerCalendarHeaderCell;
DatepickerCalendarHeaderCell.defaultProps = {
  as: 'span'
};
DatepickerCalendarHeaderCell.propTypes = _utils.commonPropTypes.createCommon();
