// @fluentui/react-northstar/dist/commonjs/components/Datepicker/DatepickerCalendarHeaderAction.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.DatepickerCalendarHeaderAction = exports.datepickerCalendarHeaderActionClassName = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactBindings = require("@fluentui/react-bindings");

var _reactIconsNorthstar = require("@fluentui/react-icons-northstar");

var _Button = require("../Button/Button");

var datepickerCalendarHeaderActionClassName = 'ui-datepicker__calendarheaderaction';
/**
 * A DatepickerCalendarHeaderAction is used to display action button for DatepickerCalendarHeader.
 * This component is currently UNSTABLE!
 */

exports.datepickerCalendarHeaderActionClassName = datepickerCalendarHeaderActionClassName;
var DatepickerCalendarHeaderAction = (0, _reactBindings.compose)(_Button.Button, {
  className: datepickerCalendarHeaderActionClassName,
  displayName: 'DatepickerCalendarHeaderAction',
  handledProps: ['direction'],
  mapPropsToStylesProps: function mapPropsToStylesProps() {
    return {
      iconOnly: true,
      text: true
    };
  },
  slotProps: function slotProps(props) {
    return {
      icon: {
        content: props.direction === 'next' ? /*#__PURE__*/React.createElement(_reactIconsNorthstar.ChevronEndIcon, null) : /*#__PURE__*/React.createElement(_reactIconsNorthstar.ChevronStartIcon, null)
      }
    };
  }
});
exports.DatepickerCalendarHeaderAction = DatepickerCalendarHeaderAction;
