// @pnp/spfx-controls-react/lib/services/SPService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { LibsOrderBy } from "./ISPService";
import { SPHttpClient } from "@microsoft/sp-http";
import { urlCombine } from "../common/utilities";
import filter from 'lodash/filter';
import find from 'lodash/find';
var SPService = /** @class */ (function () {
    function SPService(_context, webAbsoluteUrl) {
        var _this = this;
        this._context = _context;
        this.getField = function (listId, internalColumnName, webUrl) { return __awaiter(_this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, field, resultTypeRegEx, resultTypeMatch, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/fields/getByInternalNameOrTitle('" + internalColumnName + "')";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            field = results;
                            if (field.TypeAsString === 'Calculated') {
                                resultTypeRegEx = /ResultType="(\w+)"/gmi;
                                resultTypeMatch = resultTypeRegEx.exec(field.SchemaXml);
                                field.ResultType = resultTypeMatch[1];
                            }
                            return [2 /*return*/, field];
                        }
                        _a.label = 3;
                    case 3: return [3 /*break*/, 5];
                    case 4:
                        error_1 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_1)];
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        this._webAbsoluteUrl = webAbsoluteUrl ? webAbsoluteUrl : this._context.pageContext.web.absoluteUrl;
    }
    /**
     * Get lists or libraries
     *
     * @param options
     */
    SPService.prototype.getLibs = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var filtered, queryUrl, data, result, filteredLists;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        queryUrl = this._webAbsoluteUrl + "/_api/web/lists?$select=Title,id,BaseTemplate";
                        if (options.contentTypeId) {
                            queryUrl += ",ContentTypes/Id&$expand=ContentTypes";
                        }
                        if (options.orderBy) {
                            queryUrl += "&$orderby=" + (options.orderBy === LibsOrderBy.Id ? 'Id' : 'Title');
                        }
                        if (options.filter) {
                            queryUrl += "&$filter=" + encodeURIComponent(options.filter);
                        }
                        else {
                            if (options.baseTemplate) {
                                queryUrl += "&$filter=BaseTemplate eq " + options.baseTemplate;
                                filtered = true;
                            }
                            if (options.includeHidden === false) {
                                queryUrl += filtered ? ' and Hidden eq false' : '&$filter=Hidden eq false';
                                filtered = true;
                            }
                        }
                        return [4 /*yield*/, this._context.spHttpClient.get(queryUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        result = _a.sent();
                        if (options.contentTypeId) {
                            filteredLists = filter(result.value, function (aList) {
                                return find(aList.ContentTypes, function (ct) {
                                    return ct.Id.StringValue.toUpperCase().startsWith(options.contentTypeId.toUpperCase());
                                });
                            });
                            result.value = filteredLists;
                        }
                        return [2 /*return*/, result];
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Get List Items
     */
    SPService.prototype.getListItems = function (filterText, listId, internalColumnName, field, keyInternalColumnName, webUrl, filterString, substringSearch, orderBy) {
        if (substringSearch === void 0) { substringSearch = false; }
        return __awaiter(this, void 0, void 0, function () {
            var returnItems, webAbsoluteUrl, apiUrl, isPost, orderByStr, orderByParts, ascStr, camlQuery, filterStr, data, _a, results, error_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = '';
                        isPost = false;
                        if (field && field.TypeAsString === 'Calculated') { // for calculated fields we need to use CAML query
                            orderByStr = '';
                            if (orderBy) {
                                orderByParts = orderBy.split(' ');
                                ascStr = '';
                                if (orderByParts[1] && orderByParts[1].toLowerCase() === 'desc') {
                                    ascStr = "Ascending=\"FALSE\"";
                                }
                                orderByStr = "<OrderBy><FieldRef Name=\"" + orderByParts[0] + "\" " + ascStr + " />";
                            }
                            camlQuery = "<View><Query><Where>" + (substringSearch ? '<Contains>' : '<BeginsWith>') + "<FieldRef Name=\"" + internalColumnName + "\"/><Value Type=\"" + field.ResultType + "\">" + filterText + "</Value>" + (substringSearch ? '</Contains>' : '</BeginsWith>') + "</Where>" + orderByStr + "</Query></View>";
                            apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/GetItems(query=@v1)?$select=" + (keyInternalColumnName || 'Id') + "," + internalColumnName + "&@v1=" + JSON.stringify({ ViewXml: camlQuery });
                            isPost = true;
                        }
                        else {
                            filterStr = substringSearch ? // JJ - 20200613 - find by substring as an option
                                "" + (filterText ? "substringof('" + encodeURIComponent(filterText.replace("'", "''")) + "'," + internalColumnName + ")" : '') + (filterString ? (filterText ? ' and ' : '') + filterString : '')
                                : "" + (filterText ? "startswith(" + internalColumnName + ",'" + encodeURIComponent(filterText.replace("'", "''")) + "')" : '') + (filterString ? (filterText ? ' and ' : '') + filterString : '');
                            apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items?$select=" + (keyInternalColumnName || 'Id') + "," + internalColumnName + "&$filter=" + filterStr + "&$orderby=" + orderBy;
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 8, , 9]);
                        if (!isPost) return [3 /*break*/, 3];
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, {})];
                    case 2:
                        _a = _b.sent();
                        return [3 /*break*/, 5];
                    case 3: return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 4:
                        _a = _b.sent();
                        _b.label = 5;
                    case 5:
                        data = _a;
                        if (!data.ok) return [3 /*break*/, 7];
                        return [4 /*yield*/, data.json()];
                    case 6:
                        results = _b.sent();
                        if (results && results.value && results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _b.label = 7;
                    case 7: return [2 /*return*/, []];
                    case 8:
                        error_2 = _b.sent();
                        return [2 /*return*/, Promise.reject(error_2)];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    /**
  * Gets list items for list item picker
  * @param filterText
  * @param listId
  * @param internalColumnName
  * @param [keyInternalColumnName]
  * @param [webUrl]
  * @param [filterList]
  * @returns list items for list item picker
  */
    SPService.prototype.getListItemsForListItemPicker = function (filterText, listId, internalColumnName, keyInternalColumnName, webUrl, filterList) {
        return __awaiter(this, void 0, void 0, function () {
            var _filter, costumfilter, _top, webAbsoluteUrl, apiUrl, data, results, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _filter = "$filter=startswith(" + internalColumnName + ",'" + encodeURIComponent(filterText.replace("'", "''")) + "') ";
                        costumfilter = filterList
                            ? "and " + filterList
                            : "";
                        _top = " &$top=2000";
                        // test wild character "*"  if "*" load first 30 items
                        if ((filterText.trim().indexOf("*") == 0 &&
                            filterText.trim().length == 1) ||
                            filterText.trim().length == 0) {
                            _filter = "";
                            costumfilter = filterList ? "$filter=" + filterList + "&" : "";
                            _top = "$top=500";
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 5, , 6]);
                        webAbsoluteUrl = !webUrl
                            ? this._webAbsoluteUrl
                            : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists('" + listId + "')/items?$orderby=" + internalColumnName + "&$select=" + (keyInternalColumnName ||
                            "Id") + "," + internalColumnName + "&" + _filter + costumfilter + _top;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results &&
                            results.value &&
                            results.value.length > 0) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, []];
                    case 5:
                        error_3 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_3)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get list item attachments
     *
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.getListItemAttachments = function (listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId));
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results && results.value) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_4 = _a.sent();
                        console.dir(error_4);
                        return [2 /*return*/, Promise.reject(error_4)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Delete attachment
     *
     * @param fileName
     * @param listId
     * @param itemId
     * @param webUrl
     */
    SPService.prototype.deleteAttachment = function (fileName, listId, itemId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var spOpts, webAbsoluteUrl, apiUrl, data, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        spOpts = {
                            headers: { "X-HTTP-Method": 'DELETE', }
                        };
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/getByFileName(@fileName)/RecycleObject?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, spOpts)];
                    case 1:
                        data = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_5 = _a.sent();
                        console.dir(error_5);
                        return [2 /*return*/, Promise.reject(error_5)];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Add attachment
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param file
     * @param webUrl
     */
    SPService.prototype.addAttachment = function (listId, itemId, fileName, file, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var fileExists, spOpts, webAbsoluteUrl, apiUrl, data, error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        // Remove special characters in FileName
                        //Updating the escape characters for filename as per the doucmentations
                        //https://support.microsoft.com/en-us/kb/905231
                        fileName = fileName.replace(/[\~\#\%\&\*\{\}\\\:\<\>\?\/\+\|]/gi, '');
                        return [4 /*yield*/, this.checkAttachmentExists(listId, itemId, fileName, webUrl)];
                    case 1:
                        fileExists = _a.sent();
                        if (!fileExists) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.deleteAttachment(fileName, listId, itemId, webUrl)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        spOpts = {
                            body: file
                        };
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/add(FileName=@fileName)?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, spOpts)];
                    case 4:
                        data = _a.sent();
                        return [2 /*return*/];
                    case 5:
                        error_6 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_6)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get attachement for list item
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.getAttachment = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, file;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(@itemId)/AttachmentFiles/GetByFileBame(@fileName))?@listId=guid'" + encodeURIComponent(listId) + "'&@itemId=" + encodeURIComponent(String(itemId)) + "&@fileName='" + encodeURIComponent(fileName.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        file = _a.sent();
                        if (file) {
                            return [2 /*return*/, file];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                }
            });
        });
    };
    /**
     * Check if the attachment exists
     *
     * @param listId
     * @param itemId
     * @param fileName
     * @param webUrl
     */
    SPService.prototype.checkAttachmentExists = function (listId, itemId, fileName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var listServerRelativeUrl, webAbsoluteUrl, fileServerRelativeUrl, apiUrl, data, results, error_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, this.getListServerRelativeUrl(listId, webUrl)];
                    case 1:
                        listServerRelativeUrl = _a.sent();
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        fileServerRelativeUrl = listServerRelativeUrl + "/Attachments/" + itemId + "/" + fileName;
                        apiUrl = webAbsoluteUrl + "/_api/web/getfilebyserverrelativeurl(@url)/Exists?@url='" + encodeURIComponent(fileServerRelativeUrl.replace(/'/g, "''")) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 2:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 4];
                        return [4 /*yield*/, data.json()];
                    case 3:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 4;
                    case 4: return [2 /*return*/, false];
                    case 5:
                        error_7 = _a.sent();
                        return [2 /*return*/, Promise.reject(error_7)];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the list name
     *
     * @param listId
     * @param webUrl
     */
    SPService.prototype.getListName = function (listId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/RootFolder/Name?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get the list server relative url
     *
     * @param listId
     * @param webUrl
     */
    SPService.prototype.getListServerRelativeUrl = function (listId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._webAbsoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/RootFolder/ServerRelativeUrl?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.value];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getLookupValue = function (listId, listItemID, fieldName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, result, error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(" + listItemID + ")/?@listId=guid'" + encodeURIComponent(listId) + "'&$select=" + fieldName + "/ID," + fieldName + "/Title&$expand=" + fieldName;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        result = _a.sent();
                        if (result && result[fieldName]) {
                            return [2 /*return*/, [{ key: result[fieldName].ID, name: result[fieldName].Title }]];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_8 = _a.sent();
                        console.dir(error_8);
                        return [2 /*return*/, Promise.reject(error_8)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getLookupValues = function (listId, listItemID, fieldName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, result, lookups_1, error_9;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(" + listItemID + ")?@listId=guid'" + encodeURIComponent(listId) + "'&$select=" + fieldName + "/ID," + fieldName + "/Title&$expand=" + fieldName;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        result = _a.sent();
                        if (result && result[fieldName]) {
                            lookups_1 = [];
                            result[fieldName].forEach(function (element) {
                                lookups_1.push({ key: element.ID, name: element.Title });
                            });
                            return [2 /*return*/, lookups_1];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_9 = _a.sent();
                        console.dir(error_9);
                        return [2 /*return*/, Promise.reject(error_9)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getTaxonomyFieldInternalName = function (listId, fieldName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/Fields/getByInternalNameOrTitle('" + fieldName + "_0')/InternalName?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_10 = _a.sent();
                        console.dir(error_10);
                        return [2 /*return*/, Promise.reject(error_10)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getUsersUPNFromFieldValue = function (listId, listItemId, fieldName, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, result, emails_1, error_11;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/items(" + listItemId + ")?@listId=guid'" + encodeURIComponent(listId) + "'&$select=" + fieldName + "/Title," + fieldName + "/Id&$expand=" + fieldName;
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        result = _a.sent();
                        if (result && result[fieldName]) {
                            emails_1 = [];
                            result[fieldName].forEach(function (element) {
                                emails_1.push(element.Id + "/" + element.Title);
                            });
                            return [2 /*return*/, emails_1];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_11 = _a.sent();
                        console.dir(error_11);
                        return [2 /*return*/, Promise.reject(error_11)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getUserUPNById = function (userId, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_12;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/getuserbyid(" + userId + ")?$select=UserPrincipalName,Title";
                        return [4 /*yield*/, this._context.spHttpClient.get(apiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, userId + "/" + results.Title];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_12 = _a.sent();
                        console.dir(error_12);
                        return [2 /*return*/, Promise.reject(error_12)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.getSingleManagedMtadataLabel = function (listId, listItemId, fieldName) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, apiUrl, data, results, error_13;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        webAbsoluteUrl = this._context.pageContext.web.absoluteUrl;
                        apiUrl = webAbsoluteUrl + "/_api/web/lists(@listId)/RenderListDataAsStream?@listId=guid'" + encodeURIComponent(listId) + "'";
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, {
                                body: JSON.stringify({
                                    parameters: {
                                        RenderOptions: 2,
                                        ViewXml: "<View Scope=\"RecursiveAll\">\n                        <ViewFields>\n                          <FieldRef Name=\"" + fieldName + "\"/>\n                        </ViewFields>\n                        <Query>\n                          <Where>\n                            <Eq>\n                              <FieldRef Name=\"ID\"/>\n                              <Value Type=\"Number\">" + listItemId + "</Value>\n                            </Eq>\n                          </Where>\n                        </Query>\n                        <RowLimit Paged=\"TRUE\">1</RowLimit>\n                      </View>"
                                    }
                                })
                            })];
                    case 1:
                        data = _a.sent();
                        if (!data.ok) return [3 /*break*/, 3];
                        return [4 /*yield*/, data.json()];
                    case 2:
                        results = _a.sent();
                        if (results) {
                            return [2 /*return*/, results.Row[0][fieldName]];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, null];
                    case 4:
                        error_13 = _a.sent();
                        console.dir(error_13);
                        return [2 /*return*/, Promise.reject(error_13)];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    SPService.prototype.uploadImage = function (listId, itemId, fileName, file, listTitle, webUrl) {
        return __awaiter(this, void 0, void 0, function () {
            var webAbsoluteUrl, listTitleValue, listApiUrl, listResponse, listJson, apiUrl, response, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        webAbsoluteUrl = !webUrl ? this._context.pageContext.web.absoluteUrl : webUrl;
                        listTitleValue = listTitle;
                        if (!!listTitle) return [3 /*break*/, 3];
                        listApiUrl = urlCombine(webAbsoluteUrl, "/_api/web/lists('" + listId + "')", false);
                        return [4 /*yield*/, this._context.spHttpClient.get(listApiUrl, SPHttpClient.configurations.v1)];
                    case 1:
                        listResponse = _a.sent();
                        return [4 /*yield*/, listResponse.json()];
                    case 2:
                        listJson = _a.sent();
                        listTitleValue = listJson.Title;
                        _a.label = 3;
                    case 3:
                        apiUrl = urlCombine(webAbsoluteUrl, "/_api/web/UploadImage(listTitle=@a1,imageName=@a2,listId=@a3,itemId=@a4)?@a1='" + listTitleValue + "'&@a2='" + fileName + "'&@a3='" + listId + "'&@a4=" + (itemId || 0), false);
                        return [4 /*yield*/, this._context.spHttpClient.post(apiUrl, SPHttpClient.configurations.v1, {
                                body: file,
                                headers: {
                                    'content-length': file.byteLength.toString()
                                }
                            })];
                    case 4:
                        response = _a.sent();
                        return [4 /*yield*/, response.json()];
                    case 5:
                        result = _a.sent();
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return SPService;
}());
export default SPService;
