// @fluentui/react-northstar/dist/dts/src/components/Datepicker/DatepickerCalendarHeaderAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_icons_northstar_1 = require("@fluentui/react-icons-northstar");
var Button_1 = require("../Button/Button");
exports.datepickerCalendarHeaderActionClassName = 'ui-datepicker__calendarheaderaction';
/**
 * A DatepickerCalendarHeaderAction is used to display action button for DatepickerCalendarHeader.
 * This component is currently UNSTABLE!
 */
exports.DatepickerCalendarHeaderAction = react_bindings_1.compose(Button_1.Button, {
    className: exports.datepickerCalendarHeaderActionClassName,
    displayName: 'DatepickerCalendarHeaderAction',
    handledProps: ['direction'],
    mapPropsToStylesProps: function () { return ({
        iconOnly: true,
        text: true,
    }); },
    slotProps: function (props) { return ({
        icon: {
            content: props.direction === 'next' ? React.createElement(react_icons_northstar_1.ChevronEndIcon, null) : React.createElement(react_icons_northstar_1.ChevronStartIcon, null),
        },
    }); },
});
